pkg/uefi/meregion.go

```diff
@@ -0,0 +1,228 @@
+	PartitionMapStart int
+	Entries           []MEPartitionEntry
+	// Metadata for extraction and recovery
+	ExtractPath string
+}
+
+// MEPartitionEntry is an entry in FTP
+type MEPartitionEntry struct {
+	Name     MEName
+	Owner    [4]byte
+	Offset   uint32
+	Length   uint32
+	Reserved [4]uint32
+}
+
+// MEName represent 4 bytes with JSON string support
+type MEName [4]byte
+
+// MarshalText converts MEName to a byte range (for JSON)
+func (n MEName) MarshalText() ([]byte, error) {
+	return bytes.TrimRight(n[:], "\x00"), nil
+}
+
+// UnmarshalText converts a byte range to MEName (for JSON)
+func (n *MEName) UnmarshalText(b []byte) error {
+	var m MEName
+	copy(m[:], b)
+	*n = m
+	if len(b) > len(m) {
+		return fmt.Errorf("Can’t unmarshal %q to MEName, %d > %d", b, len(b), len(m))
+	}
+	return nil
+}
+
+func (n MEName) String() string {
+	b, _ := n.MarshalText()
+	return string(b)
+}
+
+// FindMESignature searches for an Intel ME FTP signature
+func FindMESignature(buf []byte) (int, error) {
+	if bytes.Equal(buf[16:16+MEFTPSignatureLength], MEFTPSignature) {
+		// 16 + 4 since the descriptor starts after the signature
+		return 20, nil
+	}
+	if bytes.Equal(buf[:MEFTPSignatureLength], MEFTPSignature) {
+		// + 4 since the descriptor starts after the signature
+		return MEFTPSignatureLength, nil
+	}
+	return -1, fmt.Errorf("ME Flash Partition Table signature not found: first 20 bytes are:\n%s",
+		hex.Dump(buf[:20]))
+}
+
+// Buf returns the buffer.
+// Used mostly for things interacting with the Firmware interface.
+func (fp *MEFPT) Buf() []byte {
+	return fp.buf
+}
+
+// SetBuf sets the buffer.
+// Used mostly for things interacting with the Firmware interface.
+func (fp *MEFPT) SetBuf(buf []byte) {
+	fp.buf = buf
+}
+
+// Apply calls the visitor on the MEFPT.
+func (fp *MEFPT) Apply(v Visitor) error {
+	return v.Visit(fp)
+}
+
+// ApplyChildren calls the visitor on each child node of MEFPT.
+func (fp *MEFPT) ApplyChildren(v Visitor) error {
+	return nil
+}
+
+// NewMEFPT tries to create a MEFPT
+func NewMEFPT(buf []byte) (*MEFPT, error) {
+	o, err := FindMESignature(buf)
+	if err != nil {
+		return nil, err
+	}
+	if len(buf) < o+28 {
+		return nil, fmt.Errorf("ME section (%#x) too small for ME Flash Partition Table (%#x)", len(buf), o+28)
+	}
+	fp := &MEFPT{PartitionMapStart: o + 28}
+	r := bytes.NewReader(buf[o:])
+	if err := binary.Read(r, binary.LittleEndian, &fp.PartitionCount); err != nil {
+		return nil, err
+	}
+	l := fp.PartitionMapStart + MEPartitionTableEntryLength*int(fp.PartitionCount)
+	if len(buf) < l {
+		return nil, fmt.Errorf("ME section (%#x) too small for %d entries in ME Flash Partition Table (%#x)", len(buf), fp.PartitionCount, l)
+	}
+
+	fp.buf = make([]byte, l)
+	copy(fp.buf, buf[:l])
+	if err := fp.parsePartitions(); err != nil {
+		return nil, err
+	}
+	return fp, nil
+}
+
+func (fp *MEFPT) parsePartitions() error {
+	r := bytes.NewReader(fp.buf[fp.PartitionMapStart:])
+	for i := 0; i < int(fp.PartitionCount); i++ {
```

Comment:
I'm pretty sure you don't need the for loop. You can do something like this: https://play.golang.org/p/UXKLQ89-zZr

Reply:
Wow ! So cool !

I'm still doing go thinking too much like C :wink: